app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java

```diff
@@ -142,19 +140,27 @@ public void onReceive(Context context, Intent intent) {
     public void updateShouldRunDecision() {
         // Check if the current conditions changed the result of decideShouldRun()
         // compared to the last determined result.
-        boolean newShouldRun = decideShouldRun();
-        if (newShouldRun != lastDeterminedShouldRun) {
+        Set<String> blockReasons = decideShouldRun();
+        boolean change;
+        synchronized (this) {
+            change = lastBlockerReasons == null || blockReasons.size() != lastBlockerReasons.size() || !lastBlockerReasons.containsAll(blockReasons);
+            lastBlockerReasons = lastBlockerReasons != null ? lastBlockerReasons : new TreeSet<>();
+        }
+        if (change) {
             if (mOnRunConditionChangedListener != null) {
-                mOnRunConditionChangedListener.onRunConditionChanged(newShouldRun);
+                mOnRunConditionChangedListener.onRunConditionChanged(blockReasons);
+            }
+            synchronized (this) {
+                lastBlockerReasons.clear();
+                lastBlockerReasons.addAll(blockReasons);
             }
-            lastDeterminedShouldRun = newShouldRun;
         }
     }
 
     /**
      * Determines if Syncthing should currently run.
      */
-    private boolean decideShouldRun() {
+    private Set<String> decideShouldRun() {
```

Comment:
Let's create a auxilary class / DTO / POJO whatever the hell, and potentially add reasons for running and a boolean?
I guess reasons can be both positive and negative, so we could store both and imply that the top one is the reason for our current state, which then allows people to understand the order of positives and negatives.

Reply:
I considered this the first time, but since the run logic had some issues previously I just didn't want to change too much to it. At the moment it aims to just be the simplest solution to the request without actually changing the logic. Future improvements like you said would be sensible, I just didn't want to do it as first step.

Reply:
Well the DTO would only have the stack of messages, that doesn't change the logic, and I think that's still valuable to understand where the issues are (when reading the stack of messages people post)